lib/Transforms/IPO/StructFieldCacheAnalysisImpl.h

```diff
@@ -681,6 +708,152 @@ class CloseProximityBuilder {
   void compareCloseProximityRelations() const;
 }; // end of class CloseProximityBuilder
 
+class StructTransformAnalyzer {
+public:
+  struct FieldDebugInfo {
+    FieldDebugInfo(FieldNumType F)
+        : FieldName(""), FieldType(""), FieldNum(F) {}
+    FieldDebugInfo(FieldNumType F, StringRef FN, StringRef FT)
+        : FieldName(FN), FieldType(FT), FieldNum(F) {}
+    StringRef FieldName;
+    StringRef FieldType;
+    FieldNumType FieldNum;
+  };
+  StructTransformAnalyzer(const Module &CM, const StructType *ST,
+                          const CloseProximityBuilder *CPB,
+                          const DICompositeType *DI);
+  ~StructTransformAnalyzer() {
+    for (auto *FDI : FieldDI) {
+      delete FDI;
+    }
+    FieldDI.clear();
+  }
+  virtual void makeSuggestions() = 0;
+
+protected:
+  const StructType *StructureType;
+  const CloseProximityBuilder *CPBuilder;
+  FieldNumType NumElements;
+  const DICompositeType *DebugInfo;
+  std::vector<unsigned> FieldSizes;
+  /// Holds a mapping between LLVM struct fields and the fields in original
+  /// source code
+  std::vector<FieldDebugInfo *> FieldDI;
+  std::vector<std::vector<double>> CloseProximityRelations;
+  /// If the struct is eligible for this kind of transformation
+  bool Eligibility;
+  /// All remaining fields to be considered
+  std::unordered_set<FieldNumType> FieldsToTransform;
+
+protected:
+  // Map fields to the debug info, useful to give recommendation and filter out
+  // padding
+  void mapFieldsToDefinition();
+  virtual double calculateCloseProximity(FieldNumType Field1,
+                                         FieldNumType Field2) const = 0;
+}; // end of class StructTransformAnalyzer
+
+class FieldReorderAnalyzer : public StructTransformAnalyzer {
+public:
+  FieldReorderAnalyzer(const Module &CM, const StructType *ST,
+                       const CloseProximityBuilder *CPB,
+                       const DICompositeType *DI);
+  ~FieldReorderAnalyzer() {}
+
+  virtual void makeSuggestions();
+
+protected:
+  // Protected constructor that is used to initialize derived class and override
+  // the public version of constructor of this class
+  FieldReorderAnalyzer(const Module &CM, const StructType *ST,
+                       const CloseProximityBuilder *CPB,
+                       const DICompositeType *DI, bool OldType)
```

Comment:
So OldFieldReorderanalyzer and FieldOrderAnalyzer are actually two separate variants of a field reorder analyzer. The confusion comes because you have named it this way. Let's call your current FieldReorderAnalyzer as NewFieldReorderAnalyzer. If the old and new versions share enough common code, put a FieldReorderAnalyzerBase class that has the common code. If they don't share much common code, derive both from the StructTransformAnalyzer.